lib/ansible/modules/files/jsonpatch.py

```diff
@@ -0,0 +1,455 @@
+#   }
+# ]
+#
+- name: add a fourth element to the "foo" object
+  jsonpatch:
+    src: "test.json"
+    dest: "test2.json"
+    operations:
+      - op: add
+        path: "/0/foo/four"
+        value: 4
+
+- name: remove the first object in the "baz" list of fruits
+  jsonpatch:
+    src: "test.json"
+    operations:
+      - op: remove
+        path: "/2/baz/0"
+
+- name: move the "potatoes" value from the "baz" list to the "foo" object
+  jsonpatch:
+    src: "test.json"
+    backup: yes
+    operations:
+      - op: move
+        from: "/2/baz/2/bar"
+        path: "/0/foo/bar"
+
+- name: test that the "foo" object has three members
+  jsonpatch:
+    src: "test.json"
+    operations:
+      - op: test
+        path: "/0/foo/one"
+        value: 1
+      - op: test
+        path: "/0/foo/two"
+        value: 2
+      - op: test
+        path: "/0/foo/three"
+        value: 3
+'''
+
+
+RETURN = r''' # '''
+
+
+import json
+import os
+
+from ansible.module_utils import basic
+from ansible.module_utils._text import to_bytes
+
+
+def set_module_args(args):
+    """For dynamic module args (such as for testing)."""
+    args = json.dumps({'ANSIBLE_MODULE_ARGS': args})
+    basic._ANSIBLE_ARGS = to_bytes(args)
+
+
+class PathError(Exception):
+    """Raised when no valid JSON object or property is found for a given path."""
+    pass
+
+
+class PatchManager(object):
+    """Manage the Ansible portion of JSONPatcher."""
+
+    def __init__(self, module):
+        self.module = module
+
+        # validate file
+        self.src = self.module.params['src']
+        if not os.path.isfile(self.src):
+            self.module.fail_json(msg="could not find file at `%s`" % self.src)
+
+        # use 'src' as the output file, unless 'dest' is provided
+        self.outfile = self.src
+        self.dest = self.module.params.get('dest')
+        if self.dest is not None:
+            self.outfile = self.dest
+
+        try:
+            self.json_doc = open(self.src).read()
+        except IOError:
+            self.module.fail_json(msg="could not read file at `%s`" % self.src)
+
+        self.operations = self.module.params['operations']
+        try:
+            self.patcher = JSONPatcher(self.json_doc, *self.operations)
+        except Exception as e:
+            self.module.fail_json(msg=str(e))
+
+        self.do_backup = self.module.params.get('backup', False)
+        self.changed = False
+
+    def run(self):
+        self.changed = self.patcher.patch()
+        if self.changed:  # let's write the changes
+            self.write()
+        return self.changed
+
+    def backup(self):
+        """Create a backup copy of the JSON file."""
+        path = self.outfile
+        counter = 0
+        while os.path.exists("%s.bak" % path):
+            path = "%s.%s" % (path, str(counter))
+            counter += 1
+        filename = "%s.bak" % path
+
+        with open(filename, "w") as bak:
+            bak.write(open(self.outfile, "r").read())
```

Comment:
Does this preserve the modification date of the original file on the backup file?

Reply:
It does not. I'll take a look at how backups are performed in other modules.